### ISSUE TITLE
CI: Teach our Azure Pipeline to run `sparse`

### ISSUE THREAD

**Issue by dscho** [gitgitgadget/git#345]
One of the earliest open source static analyzers is `sparse`, and occasionally Ramsay Jones sends out mails on the Git mailing list that some function or other should be declared `static` because `sparse` found out that it is only used within the same file.

Let's add a job to our Azure Pipeline to also run `sparse`. There is now a build offering an Ubuntu package of `sparse` as build artifact: https://dev.azure.com/git/git/_build?definitionId=10&_a=summary (which can be consumed in the same way the `git-sdk-64-minimal` build artifact is consumed, via a PowerShell step).

See the thread around https://public-inbox.org/git/20190129014426.GC7268@google.com/ for more details.

Official documentation how to use sparse: https://kernelnewbies.org/Sparse

**Comment by sgn** [comment]
Hi @dscho,

I made a draft version for this one in https://github.com/sgn/git/commit/ae06de14e5abeabdbbb89d7ba2892d85987bd4b5

Would you mind take a look, if it's OK, I'll send to the list.

I would like to discuss off-list first, to reduce the noise.

**Comment by dscho** [comment]
> I would like to discuss off-list first, to reduce the noise.

I added some comments.

Please note, however, that I am very uncomfortable with doing this "off-list". In the least, I would like to discuss this in a proper PR, so that others can see what I suggested later without having to jump through hoops.

**Comment by sgn** [comment]
I created the PR here https://github.com/git/git/pull/774

I replaced the call to `json_pp` with `jq`.
It's available already in both Travis and Github Action, and it's easier to read.

**Comment by dscho** [comment]
Since the subject [came up on the Git mailing list again](https://lore.kernel.org/git/43aadae9-a4c1-f1e4-6902-c05af36fcf19@ramsayjones.plus.com/): the Bash step to download and install the `sparse` package would look like this:

```sh
urlbase=https://dev.azure.com/git/git/_apis/build/builds
id=$(curl "$urlbase?definitions=10&statusFilter=completed&resultFilter=succeeded&\$top=1" |
	json_pp |
	sed -n 's/^         "id" : \([1-9][0-9]*\).*/\1/p')
download_url="$(curl "$urlbase/$id/artifacts" |
	json_pp |
	sed -n '/^      {/{:1;N;/\n      }/b2;b1;:2;/"name" : "sparse"/{s/.*"downloadUrl" : "\([^"]*\).*/\1/p}}')"
curl -Lo sparse.zip "$download_url"
unzip sparse.zip
sudo dpkg -i sparse/*.deb
```

**Comment by phil-blain** [comment]
This should be closed as [e610596](https://github.com/git/git/commit/e61059660ca00ba69d1721b7d48653c52848bece) was merged to `master` quite a while ago.

**Comment by dscho** [comment]
Yep, years ago :grinning:
